codegen-serde/src/main/kotlin/software/amazon/smithy/rust/codegen/serde/SerializeImplGenerator.kt

```diff
@@ -0,0 +1,500 @@
+            }
+        }
+
+    private fun serdeSubmodule(shape: Shape) =
+        RustModule.pubCrate(codegenContext.symbolProvider.shapeModuleName(codegenContext.serviceShape, shape), parent = SerdeModule)
+
+    /**
+     * Serialize a type that already implements `Serialize` directly via `value.serialize(serializer)`
+     * For enums, it adds `as_str()` to convert it into a string directly.
+     */
+    private fun directSerde(shape: Shape): RuntimeType {
+        return RuntimeType.forInlineFun(
+            codegenContext.symbolProvider.toSymbol(shape).rustType().toString(),
+            PrimitiveShapesModule,
+        ) {
+            implSerializeConfigured(codegenContext.symbolProvider.toSymbol(shape)) {
+                val baseValue =
+                    writable { rust("self.value") }.letIf(shape.isStringShape) { it.plus(writable(".as_str()")) }
+                rustTemplate("#{base}.serialize(serializer)", "base" to baseValue)
+            }
+        }
+    }
+
+    /**
+     * Serialize a shape by first generating a wrapper struct:
+     * ```rust
+     * struct WrapperType<'a>(&'a Type);
+     * ```
+     *
+     * Then implementing `Serialize` for `ConfigurableSerdeRef<'a WrapperType>`
+     *
+     * This exists to allow differing implementations for same-shaped-rust types. For example, consider
+     * the following Smithy model:
+     *
+     * ```smithy
+     * list SensitiveList {
+     *   member: SensitiveString
+     * }
+     *
+     * list StringList {
+     *   member: String
+     * }
+     * ```
+     *
+     * These both are `Vec<String>` but must be serialized differently.
+     */
+    private fun serializeWithWrapper(
+        shape: Shape,
+        body: (Writable) -> Writable,
+    ): RuntimeType {
+        val name =
+            "Serializer" +
+                codegenContext.symbolProvider.shapeFunctionName(codegenContext.serviceShape, shape)
+                    .toPascalCase()
+        // awkward hack to recover the symbol referring to the type
+        val module = serdeSubmodule(shape)
+        val type = module.toType().resolve(name).toSymbol()
+        val base =
+            writable { rust("self.value.0") }.letIf(shape.hasConstraintTrait() && codegenContext.target == CodegenTarget.SERVER) {
+                it.plus { rust(".0") }
+            }
+        val serialization =
+            implSerializeConfiguredWrapper(type) {
+                body(base)(this)
+            }
+        val wrapperStruct =
+            RuntimeType.forInlineFun(name, module) {
+                rustTemplate(
+                    """
+                    pub(crate) struct $name<'a>(pub(crate) &'a #{Shape});
+                    #{serializer}
+                    """,
+                    "Shape" to codegenContext.symbolProvider.toSymbol(shape),
+                    "serializer" to serialization,
+                )
+            }
+        return wrapperStruct
+    }
+
+    private fun Shape.unwrapConstraints(): Boolean =
```

Comment:
I think you'll also need to take into account whether the `publicConstrainedTypes` setting is disabled. If it is then the user is not exposed to the wrapper newtypes and it doesn't matter whether the shapes are constrained or not, they'll work with types generated as if the constraints hadn't been modeled.

Reply:
done